scipy/sparse/_coo.py

```diff
@@ -722,16 +810,463 @@ def _matmul_multivector(self, other):
+        other_broadcasted = other._broadcast_to(other_new_shape)
+        
+        # Convert n-D COO arrays to 2-D block diagonal arrays
+        self_block_diag = _block_diag(self_broadcasted)
+        other_block_diag = _block_diag(other_broadcasted)
+        
+        # Use csr_matmat to perform sparse matrix multiplication
+        prod_block_diag = (self_block_diag @ other_block_diag).tocoo()
+        
+        # Convert the 2-D block diagonal array back to n-D
+        prod = _extract_block_diag(prod_block_diag, shape=(*broadcast_shape,
+                                                     self.shape[-2], other.shape[-1]))
+        
+        return prod
+
+
+    def _broadcast_to(self, new_shape, copy=False):
+        if self.shape == new_shape:
+            return self.copy() if copy else self
+        
+        old_shape = self.shape
+
+        # Check if the new shape is compatible for broadcasting
+        if len(new_shape) < len(old_shape):
+            raise ValueError("New shape must have at least as many dimensions"
+                             " as the current shape")
+        
+        # Add leading ones to shape to ensure same length as `new_shape` 
+        shape = (1,) * (len(new_shape) - len(old_shape)) + tuple(old_shape)
+        
+        # Ensure the old shape can be broadcast to the new shape
+        if any((o != 1 and o != n) for o, n in zip(shape, new_shape)):
+            raise ValueError(f"current shape {old_shape} cannot be "
+                             "broadcast to new shape {new_shape}")
+
+        # Reshape the COO array to match the new dimensions
+        self = self.reshape(shape)
+
+        coords = np.array(self.coords)
+        new_data = np.array(self.data) # copy data array
+        new_coords = coords[-1:]  # Copy last coordinate to start
+        cum_repeat = 1 # Cumulative repeat factor for broadcasting
+        
+        if shape[-1] != new_shape[-1]: # broadcasting the n-th (col) dimension
+            repeat_count = new_shape[-1]
+            cum_repeat *= repeat_count
+            new_data = np.tile(new_data, repeat_count)
+            new_dim = np.repeat(np.arange(0, repeat_count), self.nnz)
+            new_coords = np.array([new_dim])
+        
+        for i in range(-2, -(len(shape)+1), -1):
+            if shape[i] != new_shape[i]:
+                repeat_count = new_shape[i] # number of times to repeat data, coords
+                cum_repeat *= repeat_count # update cumulative repeat factor
+                nnz = len(new_data) # Number of non-zero elements so far
+
+                # Tile data and coordinates to match the new repeat count
+                new_data = np.tile(new_data, repeat_count)
+                new_coords = np.tile(new_coords[i+1:], repeat_count)
+
+                # Create new dimensions and stack them
+                new_dim = np.repeat(np.arange(0, repeat_count), nnz)
+                new_coords = np.vstack((new_dim, new_coords))
```

Comment:
Can we avoid the copying from `vstack` here by making `new_coords` a tuple throughout? (lines 1148, 1150, 1158, 1168, 1172, 1176)

Reply:
Done at 7c8d4980d2c7d1e1011b521168583ed2639b7b6b.
A minor issue - in line 1168, even though `np.tile` takes a tuple-of-arrays input, it returns an array-of-arrays output. This was causing addition instead of contatenation behaviour when we had 2 single-array-tuples, so I had to convert the output from tiling to a tuple. I'd assume this may be a bit inefficient.

Reply:
You are correct, but the tuple is not very long (ndim<65) and it avoids copying the `self.nnz` values. So it is still better than using concatenate.  `nx.tile` copies the data. But `tuple(nx.tile(...))` does not recopy it -- iteration produces views on the array. 

I don't know any way to avoid copying the data while tiling. But there may be ways to avoid literal broadcasting by repeatedly using the same values for the broadcast dimension. But that is not for this PR. :)